File: Seori/LeetCode 75/Hash Map_Set/2352. Equal Row and Column Pairs.py
class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        # [1] grid를 반시계방향으로 회전한다.
        n = len(grid)
        grid_counterclockwise = []
        for i in range(n):
            column = []
            for j in range(n):
                column.append(grid[j][i])
            grid_counterclockwise.append(column)
        
        # [2] 두 그리드를 비교하여 같은 개수를 센다.
        answer = 0
        for row in grid:
            for column in grid_counterclockwise:
                if row == column:
                    answer += 1

        return answer

File: Lucy/2025/W03/2352_Equal_Row_and_Column_Pairs.ts
function equalPairs(grid: number[][]): number {
  const SEPARATION = "|";
  const INIT_VALUE = 0;

  let answer = 0;
  const rowNumberCount = new Map();
  const colNumberCount = new Map();

  grid.forEach((row, i, arr) => {
    const joinedRow = row.join(SEPARATION);
    rowNumberCount.set(joinedRow, (rowNumberCount.get(joinedRow) || INIT_VALUE) + 1);

    const joinedCol = arr.map((row) => row[i]).join(SEPARATION);
    colNumberCount.set(joinedCol, (colNumberCount.get(joinedCol) || INIT_VALUE) + 1);
  });

  const rowNumbers = Array.from(rowNumberCount.keys());
  const colNumbers = Array.from(colNumberCount.keys());

  const numberKeys = new Set([...rowNumbers, ...colNumbers]);

  numberKeys.forEach((key) => {
    if (rowNumberCount.has(key) && colNumberCount.has(key)) {
      answer += rowNumberCount.get(key) * colNumberCount.get(key);
    }
  });

  return answer;
}

equalPairs([
  [3, 2, 1],
  [1, 7, 6],
  [2, 7, 7],
]); // 1

equalPairs([
  [3, 1, 2, 2],
  [1, 4, 4, 5],
  [2, 4, 2, 2],
  [2, 4, 2, 2],
]); // 3
